libnavui-androidauto/src/main/java/com/mapbox/androidauto/screencoordinator/MapboxScreenManager.kt

```diff
@@ -0,0 +1,237 @@
+        }
+
+        override fun onDestroy(owner: LifecycleOwner) {
+            onDestroy()
+        }
+
+        override fun onStart(owner: LifecycleOwner) {
+            coroutineScope = CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate)
+            coroutineScope.launch {
+                screenEvent
+                    .filterNotNull()
+                    .collect {
+                        onCarScreenEvent(it)
+                        logAndroidAuto("MapboxScreenManager screen changed $it")
+                    }
+            }
+        }
+
+        override fun onStop(owner: LifecycleOwner) {
+            coroutineScope.cancel()
+        }
+    }
+
+    /**
+     * This call is required. Many functions will throw [IllegalStateException] if this is not
+     * called. Ensure that you use this object while the [LifecycleOwner] is at least in the
+     * [Lifecycle.State.CREATED] state. The object is automatically destroyed when the
+     * [LifecycleOwner] moves to the [Lifecycle.State.DESTROYED] state.
+     */
+    @UiThread
+    fun setup(lifecycleOwner: LifecycleOwner, carContext: CarContext) = apply {
+        this.lifecycleOwner = lifecycleOwner
+        this.carContext = carContext
+        MapboxNavigationApp.attach(lifecycleOwner)
+        lifecycleOwner.lifecycle.addObserver(lifecycleObserver)
+    }
+
+    /**
+     * Gives you the ability to completely disable the [MapboxScreenManager]. This is not required
+     * for cleaning up the object because the lifecycleOwner will trigger the clean up when the
+     * lifecycle has been destroyed.
+     */
+    @UiThread
+    fun onDestroy() {
+        lifecycleOwner?.let {
+            MapboxNavigationApp.detach(it)
+            it.lifecycle.removeObserver(lifecycleObserver)
+        }
+        screenFactoryMap.clear()
+        coroutineScope.cancel()
+        carContext = null
+        screenManager = null
+        lifecycleOwner = null
+    }
+
+    /**
+     * This can be used to create a screen from [Session.onCreateScreen]. Using this function will
+     * also notify any [screenKey] observers with [MapboxScreenOperation.CREATED].
+     */
+    @UiThread
+    fun create(screenKey: String): Screen {
+        val factory: MapboxScreenFactory = requireScreenFactory(screenKey)
+        return factory.create(requireCarContext()).also {
+            screenKeyMutable.value = MapboxScreenEvent(
+                screenKey,
+                MapboxScreenOperation.CREATED
+            )
+        }
+    }
+
+    /**
+     * Returns the previously set screen factory.
+     */
+    operator fun <T : MapboxScreenFactory> set(key: String, factory: T): MapboxScreenFactory? {
+        return screenFactoryMap.put(key, factory)
+    }
+
+    /**
+     * Allows you to put all defined screen factories into the manager in one operation.
+     */
+    fun putAll(map: Map<String, MapboxScreenFactory>) {
+        return screenFactoryMap.putAll(map)
```

Comment:
I think this will make it more convenient to call this function:
```suggestion
    fun putAll(vararg pairs: Pair<String, MapboxScreenFactory>) {
        return screenFactoryMap.putAll(pairs)
```

Reply:
oh yeah thanks! i just copy pasted `public fun putAll(from: Map<out K, V>): Unit` from the kotlin `MutableMap`

love this